op-challenger/challenger/traversal.go

```diff
@@ -0,0 +1,238 @@
+package challenger
+
+import (
+	"context"
+	"errors"
+	"math/big"
+	"time"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+
+	"github.com/ethereum-optimism/optimism/op-service/backoff"
+)
+
+var (
+	ErrAlreadyStarted = errors.New("already started")
+)
+
+//go:generate mockery --name TraversalClient
+type TraversalClient interface {
+	SubscribeNewHead(ctx context.Context, ch chan<- *types.Header) (ethereum.Subscription, error)
+	TransactionReceipt(ctx context.Context, txHash common.Hash) (*types.Receipt, error)
+	BlockByHash(ctx context.Context, hash common.Hash) (*types.Block, error)
+	HeaderByNumber(ctx context.Context, number *big.Int) (*types.Header, error)
+}
+
+// LogTraversal is a client that can traverse all logs.
+type LogTraversal interface {
+	Start(context.Context, chan struct{}, func(types.Log) error) error
+	Quit()
+}
+
+// logTraversal implements LogTraversal.
+type logTraversal struct {
+	log             log.Logger
+	client          TraversalClient
+	query           *ethereum.FilterQuery
+	quit            chan struct{}
+	lastBlockNumber *big.Int
+	started         bool
+}
+
+// NewLogTraversal creates a new log traversal.
+func NewLogTraversal(client TraversalClient, query *ethereum.FilterQuery, log log.Logger) *logTraversal {
+	return &logTraversal{
+		client:          client,
+		query:           query,
+		quit:            make(chan struct{}),
+		log:             log,
+		lastBlockNumber: big.NewInt(0),
```

Comment:
I think this is going to iterate over every block from genesis every time we start up which will take days.

Reply:
ditto